### ISSUE TITLE
Rule: S6145 "Option Strict" should be enabled

### ISSUE THREAD

**Issue by pavel-mikula-sonarsource** [SonarSource/sonar-dotnet#3984]
See #2399
Originally posted by @Corniel 

--------------------
* Title: Option Strict should not be disabled
* Type: Code Smell Detection
* Labels: None
* Message: Option Strict should not be disabled
* Default Severity: Major
* Impact: Major
* Default Quality Profiles: Sonar way	
* Targeted languages: VB.Net
* Constant Cost: ?
* Analysis Level: Syntactic Analysis	
* Analysis Scope: Main Sources, Test Sources	

# VB.NET
Description:
`Option Strict` has been introduced to not break VB6 legacy code, and still be able to disallow:
* Implicit narrowing conversions
* Late binding
* Implicit typing that results in an `Object` type

## Noncompliant Code Example
``` VB.NET
Option Strict On
```

## Exceptions
It can be used as an alternative to `dynamic` as introduced in C#.
--------------------

The value can also be set in a project file. It might be worth to start validation project files to, but that would change the scope of this (now) tiny pull request dramatically.

**Comment by pavel-mikula-sonarsource** [comment]
cc @Corniel , I moved the content here to keep the PR #3983  just a PR

**Comment by pavel-mikula-sonarsource** [comment]
Hi @Corniel ,

This is a very good rule, I'll try to process it during this sprint. We'll need to start with the RSPEC to give you rule ID and scaffolding.

Few questions:
* The `Noncompliant code example` is actually compliant right? `Option Strict Off` is the pain we're trying to prevent.
* Why do you list the `Exceptions`? It's not gonna be implemented so it shouldn't be there.

**Comment by pavel-mikula-sonarsource** [comment]
Let's extend the rule to prohibit `Option Explicit Off` as well.

**Comment by Corniel** [comment]
Sorry, my mistake: the error message was wrong: It should be *enabled*.

The reason should not disallow both is 'simple'. Lets assume someone has the following project file:

``` XML
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OptionStrict>Off</OptionStrict>
  </PropertyGroup>

</Project>
```

Than turning it on for one file is better that not turning it one at all. As mentioned in t PR, it would be nice to have project file validations too, but as there is not mechanism in place yet, that might to be the best place to start. For now, I think it would be wice to allow `Option Strict On`, and to disallow `Option Strict Off`.

**Comment by pavel-mikula-sonarsource** [comment]
There's also `Option Strict Custom` so I'd disallow `Option Strict Off` in the implementation. The message could ask for allowing it.

It seems to be easy to check the project file configuration as well - with registration like this https://github.com/SonarSource/sonar-dotnet/blob/65309d7aa8580b7186178b147b6d9913a19e278b/analyzers/src/SonarAnalyzer.CSharp/Rules/MarkAssemblyWithAttributeBase.cs#L30-L38 
you can inspect `((VisualBasicCompilationOptions)c.Compilation.Options).OptionStrict` and report on assembly level. The assembly name should be then part of the message in that case, as it's not reported on any location.

**Comment by Corniel** [comment]
Have some doubts about the messages to show, but validation should be in place.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @Corniel, this rule is not specified as [S6145](https://jira.sonarsource.com/browse/RSPEC-6145). I've suggested two messages for the rule
* `Change this to 'Option Strict On'` - for syntax based detection
* `Configure 'Option Strict On' for assembly '{0}'.` - for project properties level detection

I've added scaffolding for you in https://github.com/SonarSource/sonar-dotnet/tree/Pavel/Corniel-S6145-Scaffolding branch

I've also specified it's sibling [S6146](https://jira.sonarsource.com/browse/RSPEC-6146) for `Option Explicit`
